### ISSUE TITLE
Infinite prompts "file changed" with revert-buffer-insert-file-contents-delicately + Evil combo

### ISSUE THREAD

**Issue by Hi-Angel** [emacs-evil/evil#1504]
# Issue type

- Bug report

# Environment

Emacs version: GNU Emacs 28.0.50 (build 4, x86_64-pc-linux-gnu, GTK+ Version 3.24.30, cairo version 1.17.4) of 2021-08-03
Operating System: Archlinux
Evil version: Evil version 1.14.0
Evil installation type: MELPA
Graphical/Terminal: Graphical
Tested in a `make emacs` session (see [CONTRIBUTING.md](https://github.com/emacs-evil/evil/blob/master/CONTRIBUTING.md#bug-reports)): Yes

# Reproduction steps
1. Start Emacs with `make emacs` from evil repo
2. Evaluate `(setq revert-buffer-insert-file-contents-function 'revert-buffer-insert-file-contents-delicately)`
3. Evaluate `(global-auto-revert-mode 1)`
4. Execute `echo hello > /tmp/1`
5. In the Emacs, open the created `/tmp/1` file
6. Press <kbd>G</kbd> to cause initial position to be put into evil jumps buffer.
7. Execute `echo hello > /tmp/1`
8. Press <kbd>C-o</kbd> to jump back

# Expected behavior

Caret changes its position

# Actual behavior

Emacs asks `File 1 changed on disk.  Reread from disk? (yes or no)`

# Further notes

I am not sure if it's upstream bug or Evil, but as I presume Evil might be using its own buffer to keep jump positions, so I decided to report it here. Please tell if it needs to be moved upstream *(and why, so I can create the report)*.

The prompt will be popping up infinitely on every move with <kbd>C-o</kbd> and <kbd>C-i</kbd> until you make a change in the buffer.

`2` is the specific step causing the problem. This step is a workaround for Emacs breaking markers in buffers upon reverting them, [see also this report](https://github.com/magit/magit/issues/4442). However, attempt to use the workaround turned up this even worse problem than breaking markers in a buffer.

**Comment by Hi-Angel** [comment]
> I am not sure if it's upstream bug or Evil, but as I presume Evil might be using its own buffer to keep jump positions, so I decided to report it here. Please tell if it needs to be moved upstream *(and why, so I can create the report)*.

On a second thought, IIUC `revert-buffer-insert-file-contents-delicately` should behave similarly to `revert-buffer-insert-file-contents--default-function`, and since Evil behaves differently, something is wrong upstream.

Still, I'd appreciate hints from people knowing Evil code on what exactly goes wrong, so I could craft up a testcase that would not involve installing Evil.

**Comment by Hi-Angel** [comment]
> Still, I'd appreciate hints from people knowing Evil code on what exactly goes wrong, so I could craft up a testcase that would not involve installing Evil.

I managed to reduce the bug. What happens is that upon calling `evil-jump-backward` and `evil-jump-forward`, Evil always calls `(find-file file-name)`. So, by changing the testcase to call `find-file` I'm able to reproduce the problem without Evil.

I will be reporting the problem upstream once I gather a bit more info, but I found there is inefficiency in Evil implementation as well: it should not call `find-file` nor `switch-to-buffer` when the current buffer has a file associated and the jump target is that file *(which is the most frequent situation)*. I will sent an improvement to that later this evening as well.
